reco_utils/recommender/ncf/dataset.py

```diff
@@ -0,0 +1,290 @@
+                    self.test_ratings.append(float(0))
+
+                self.test_data.append ( [
+                        [self.id2user[x] for x in self.test_users],
+                        [self.id2item[x] for x in self.test_items],
+                        self.test_ratings
+                    ]
+                )
+
+    def negative_sampling(self):
+        """ sample n_neg negative items per positive item
+            this function should be called every epoch
+        """
+        self.users, self.items, self.ratings = [], [], []
+
+        # sample n_neg negative samples for training
+        train_ratings = pd.merge(self.train, self.interact_status[[self.col_user, self.col_item + '_negative']],
+                                 on=self.col_user)
+
+        try:
+            train_ratings[self.col_item + '_negative'] = train_ratings[self.col_item + '_negative'].apply(
+                lambda x: random.sample(x, self.n_neg))
+        except:
+            min_num = min(map(len, list(train_ratings[self.col_item + '_negative'])))
+            warnings.warn("n_neg is larger than negative items set size! We will set n_neg as the smallest size: %d" % min_num)
+            train_ratings[self.col_item + '_negative'] = train_ratings[self.col_item + '_negative'].apply(
+                lambda x: random.sample(x, min_num)
+            )
+
+        # generate training data
+        for row in train_ratings.itertuples():
+            self.users.append(int(getattr(row, self.col_user)))
+            self.items.append(int(getattr(row, self.col_item)))
+            self.ratings.append(float(getattr(row, self.col_rating)))
+            for i in getattr(row, self.col_item + '_negative'):
+                self.users.append(int(getattr(row, self.col_user)))
+                self.items.append(int(i))
+                self.ratings.append(float(0))
+
+        self.users = np.array(self.users)
+        self.items = np.array(self.items)
+        self.ratings = np.array(self.ratings)
+
+    def train_loader(self, batch_size, shuffle=True):
+        """ feed train data every batch
+            Args:
+                batch size (int)
+                shuffle (bool): if true, train data will be shuffled
+            Returns:
+                list: userID list, itemID list, rating list.
+                public data loader return the userID, itemID consistent with raw data
+
+        """
+
+        # yield batch of training data with `shuffle`
+
+        indices = np.arange(len(self.users))
+        if shuffle:
+            np.random.shuffle(indices)
+        for i in range(len(indices) // batch_size):
+            begin_idx = i * batch_size
+            end_idx = (i + 1) * batch_size
+            batch_indices = indices[begin_idx: end_idx]
+            yield [
+                [self.id2user[x] for x in self.users[batch_indices]],
+                [self.id2item[x] for x in self.items[batch_indices]],
+                self.ratings[batch_indices],
```

Comment:
it appears as though when this data is used it is converted back to ids? is that right? can we just keep it as ideas and pass it to TF?

Reply:
@gramhagen Yes, actually in the original version, we pass ids to TF directly. After discussing with @yueguoguo, we think function `train_loader()` could be called and used by our users in other situations , they expect the not re-indexed data. So we convert id --> user --> id to TF when using data loader. Yet that's slower.

Reply:
I see, ok